airflow/providers/cncf/kubernetes/executors/kubernetes_executor.py

```diff
@@ -577,6 +577,17 @@ def try_adopt_task_instances(self, tis: Sequence[TaskInstance]) -> Sequence[Task
                 for pod in pod_list:
                     self.adopt_launched_task(kube_client, pod, tis_to_flush_by_key)
             self._adopt_completed_pods(kube_client)
+            # as this method can be retried within a short time frame
+            # (wrapped in a run_with_db_retries of scheduler_job_runner,
+            # and get retried due to an OperationalError, for example),
+            # there is a chance that in second attempt, adopt_launched_task will not be called even once
+            # as all pods are already adopted in the first attempt.
+            # and tis_to_flush_by_key will contain TIs that are already adopted.
+            # therefore, we need to check if the TIs are already adopted by the first attempt and remove them.
+            for ti in list(tis_to_flush_by_key.keys()):
+                if ti in self.running:
+                    del tis_to_flush_by_key[ti]
+                    self.log.info("%s is already adopted, no need to flush.", ti)
             tis_to_flush.extend(tis_to_flush_by_key.values())
```

Comment:
Instead of trying to modify `tis_to_flush_by_key` in-place, I would just do

```python
for key, ti in tis_to_flush_by_key.items():
    if key in self.running:
        self.log.info("%s is already adopted, no need to flush.", ti)
    else:
        tis_to_flush.append(ti)
```

or even better:

```python
def _iter_tis_to_flush():
    for key, ti in tis_to_flush_by_key.items():
        if key in self.running:
            self.log.info("%s is already adopted, no need to flush.", ti)
        else:
            yield ti

tis_to_flush.extend(_iter_tis_to_flush())
```

Reply:
Thanks and fixed!